### ISSUE TITLE
TreeEnsembleClassifier with post_transform=LOGISTIC skips sigmoid for leaf-only trees when all weights are non-negative

### ISSUE THREAD

**Issue by hopper-signifyd** [microsoft/onnxruntime#27533]
## Describe the issue

We have code which converts xgboost models to ONNX. We were seeing major score mismatches when attempting to upgrade from xgboost 1.7.2 to xgboost 3. With the help of AI, I cloned down the source code for the various OS repos involved. AI believes the issue lives in the onnxruntime itself and was the result of our training dataset being too small to produce splits. To work around this, we just increased the size of our test dataset. But I figured it's worth opening a bug report about this edge case. The AI generated bug report is as follows:

`TreeEnsembleClassifier` with `post_transform=LOGISTIC` does not apply the sigmoid transformation when **all tree weights are non-negative** (`weights_are_all_positive_ = true`). This happens for binary classifiers where every tree is a single leaf node (no splits), which is a valid degenerate case produced by XGBoost when training data is too small to learn splits.

**Expected behavior**: The second output (class scores) should contain post-transformed probabilities: `[1 - sigmoid(agg), sigmoid(agg)]`

**Actual behavior**: The second output contains raw scores without sigmoid: `[1 - agg, agg]`

The bug is in the interaction between `_set_score_binary` in `tree_ensemble_aggregator.h` and `write_scores` in `ml_common.h`. When `weights_are_all_positive_ == true`, `_set_score_binary` sets `add_second_class` to 0 or 1. The `write_scores` function only applies LOGISTIC (sigmoid) for `add_second_class` values 2 and 3 — values 0 and 1 output `[1-score, score]` without sigmoid.

Trees with real splits (which produce both positive and negative weights) correctly set `weights_are_all_positive_ = false`, causing `add_second_class = 2 or 3`, and the sigmoid IS applied. So the bug only manifests for leaf-only trees.

### Source code references

1. **`tree_ensemble_aggregator.h` — `_set_score_binary()`**: When `weights_are_all_positive_ == true`, overwrites `write_additional_scores` to 0 or 1:
   ```cpp
   if (weights_are_all_positive_) {
       if (pos_weight > 0.5) {
           write_additional_scores = 0;  // <-- bypasses sigmoid
           return class_labels_[1];
       } else {
           write_additional_scores = 1;  // <-- bypasses sigmoid
           return class_labels_[0];
       }
   }
   ```

2. **`ml_common.h` — `write_scores()`**: Only applies LOGISTIC for `add_second_class` 2 and 3:
   ```cpp
   switch (add_second_class) {
       case 0:
       case 1:
           // Raw score output — NO sigmoid applied
           scores.push_back(scores[0]);
           scores[0] = 1 - scores[0];
           break;
       case 2:
       case 3:
           if (post_transform == POST_EVAL_TRANSFORM::LOGISTIC) {
               // Sigmoid IS applied here
               scores[1] = ComputeLogistic(scores[0]);
               scores[0] = ComputeLogistic(-scores[0]);
           }
           break;
   }
   ```

3. **`tree_ensemble_common.h`**: `weights_are_all_positive_` is set to `true` when all `class_weights` values are non-negative. Leaf-only XGBoost binary:logistic trees always have non-negative weights.

## To reproduce

Minimal self-contained reproducer:

```python
import numpy as np
import onnxruntime
from onnx import TensorProto, helper

print(f"onnxruntime version: {onnxruntime.__version__}")

X = helper.make_tensor_value_info("X", TensorProto.FLOAT, [None, 3])
label_out = helper.make_tensor_value_info("label", TensorProto.INT64, [None])
prob_out = helper.make_tensor_value_info("probs", TensorProto.FLOAT, [None, 2])


def make_model(nodes_modes, nodes_values, nodes_truenodeids, nodes_falsenodeids,
               class_treeids, class_nodeids, class_weights, **node_kwargs):
    """Build a minimal TreeEnsembleClassifier ONNX model."""
    n_nodes = len(nodes_modes)
    node = helper.make_node(
        "TreeEnsembleClassifier",
        inputs=["X"],
        outputs=["label", "probs"],
        domain="ai.onnx.ml",
        nodes_treeids=[0] * n_nodes,
        nodes_nodeids=list(range(n_nodes)),
        nodes_featureids=[0] * n_nodes,
        nodes_values=nodes_values,
        nodes_modes=nodes_modes,
        nodes_truenodeids=nodes_truenodeids,
        nodes_falsenodeids=nodes_falsenodeids,
        nodes_missing_value_tracks_true=[0] * n_nodes,
        nodes_hitrates=[1.0] * n_nodes,
        class_treeids=class_treeids,
        class_nodeids=class_nodeids,
        class_ids=[0] * len(class_weights),
        class_weights=class_weights,
        classlabels_int64s=[0, 1],
        base_values=[-0.405],  # logit(0.4)
        post_transform="LOGISTIC",
        **node_kwargs,
    )
    graph = helper.make_graph([node], "test", [X], [label_out, prob_out])
    return helper.make_model(graph, opset_imports=[
        helper.make_opsetid("", 15),
        helper.make_opsetid("ai.onnx.ml", 3),
    ])


test_input = {"X": np.array([[0.1, 0.0, 0.0]], dtype=np.float32)}

# Case 1: Tree with a real split (root splits on feature 0 at 0.5)
# Weights include both positive and negative values -> weights_are_all_positive_ = false
model_split = make_model(
    nodes_modes=["BRANCH_LT", "LEAF", "LEAF"],
    nodes_values=[0.5, 0.0, 0.0],
    nodes_truenodeids=[1, 0, 0],
    nodes_falsenodeids=[2, 0, 0],
    class_treeids=[0, 0],
    class_nodeids=[1, 2],
    class_weights=[0.3, -0.3],  # mixed positive/negative
)
sess_split = onnxruntime.InferenceSession(model_split.SerializeToString())
result_split = sess_split.run(None, test_input)
# x[0]=0.1 < 0.5, so left leaf (weight=0.3), aggregate = -0.405 + 0.3 = -0.105
expected_p1 = 1 / (1 + np.exp(0.105))  # sigmoid(-0.105)
print(f"Tree with splits:    probs={result_split[1]}")
print(f"  Expected: [[{1-expected_p1:.6f}, {expected_p1:.6f}]]")
print(f"  Sigmoid correctly applied: {np.allclose(result_split[1][0][1], expected_p1, atol=1e-5)}")
print()

# Case 2: Leaf-only tree (single LEAF node, no splits)
# All weights are non-negative -> weights_are_all_positive_ = true
model_leaf = make_model(
    nodes_modes=["LEAF"],
    nodes_values=[0.0],
    nodes_truenodeids=[0],
    nodes_falsenodeids=[0],
    class_treeids=[0],
    class_nodeids=[0],
    class_weights=[0.0],  # non-negative weight
)
sess_leaf = onnxruntime.InferenceSession(model_leaf.SerializeToString())
result_leaf = sess_leaf.run(None, test_input)
# aggregate = -0.405 + 0 = -0.405
expected_p1_leaf = 1 / (1 + np.exp(0.405))  # sigmoid(-0.405) ≈ 0.400
print(f"Leaf-only tree:      probs={result_leaf[1]}")
print(f"  Expected: [[{1-expected_p1_leaf:.6f}, {expected_p1_leaf:.6f}]]")
print(f"  Sigmoid correctly applied: {np.allclose(result_leaf[1][0][1], expected_p1_leaf, atol=1e-5)}")
print()

# Case 3: Same leaf-only tree but with a negative weight (workaround)
# weights_are_all_positive_ = false -> sigmoid IS applied
model_leaf_neg = make_model(
    nodes_modes=["LEAF"],
    nodes_values=[0.0],
    nodes_truenodeids=[0],
    nodes_falsenodeids=[0],
    class_treeids=[0],
    class_nodeids=[0],
    class_weights=[-0.405],  # negative weight (move base_values into weight)
    base_values=[0.0],       # zero base
)
sess_leaf_neg = onnxruntime.InferenceSession(model_leaf_neg.SerializeToString())
result_leaf_neg = sess_leaf_neg.run(None, test_input)
print(f"Leaf-only (neg wt):  probs={result_leaf_neg[1]}")
print(f"  Expected: [[{1-expected_p1_leaf:.6f}, {expected_p1_leaf:.6f}]]")
print(f"  Sigmoid correctly applied: {np.allclose(result_leaf_neg[1][0][1], expected_p1_leaf, atol=1e-5)}")
```

### Expected output

All three cases should produce post-transformed probabilities (sigmoid applied):
```
Tree with splits:    probs=[[0.52622586 0.47377414]]   # sigmoid applied ✓
Leaf-only tree:      probs=[[0.5998884  0.40011162]]   # sigmoid applied ✓
Leaf-only (neg wt):  probs=[[0.5998884  0.40011162]]   # sigmoid applied ✓
```

### Actual output (onnxruntime 1.24.2)

```
Tree with splits:    probs=[[0.52622586 0.47377414]]   # sigmoid applied ✓
Leaf-only tree:      probs=[[ 1.405     -0.405    ]]   # sigmoid NOT applied ✗ — raw scores
Leaf-only (neg wt):  probs=[[0.5998884  0.40011162]]   # sigmoid applied ✓ (workaround)
```

The leaf-only case returns raw `[1 - logit, logit]` instead of `[1 - sigmoid(logit), sigmoid(logit)]`. The third case shows that making the weight negative (moving the base_values into the weight) causes `weights_are_all_positive_ = false` and the sigmoid is correctly applied.

## Urgency

This bug causes incorrect probability outputs for XGBoost binary:logistic models converted to ONNX via onnxmltools. It primarily affects degenerate models where all trees are leaf-only (e.g., when training data is very small), but could also affect any model where all `class_weights` happen to be non-negative.

The `post_transform` attribute is part of the ONNX ML spec and should be honored regardless of weight sign patterns.

## Platform

Mac (also reproducible on Linux)

## OS Version

macOS 15.4 (Darwin 25.2.0)

## ONNX Runtime Installation

Released Package

## ONNX Runtime Version or Commit ID

1.24.2

## ONNX Runtime API

Python

## Architecture

ARM64

## Execution Provider

Default CPU
